src/core/reactor.cc

```diff
@@ -2075,49 +2075,34 @@ static auto next_waitpid_timeout(std::chrono::milliseconds this_timeout) {
 
 #endif
 
-future<int> reactor::do_waitpid(pid_t pid) {
-    return do_with(int{}, std::chrono::milliseconds(0), [pid, this](int& wstatus,
-                                                                    std::chrono::milliseconds& wait_timeout) {
-        return repeat_until_value([this,
-                                   pid,
-                                   &wstatus,
-                                   &wait_timeout] {
-            return _thread_pool->submit<syscall_result<pid_t>>([pid, &wstatus] {
-                return wrap_syscall<pid_t>(::waitpid(pid, &wstatus, WNOHANG));
-            }).then([&wstatus, &wait_timeout](syscall_result<pid_t> ret) mutable {
-                if (ret.result == 0) {
-                    wait_timeout = next_waitpid_timeout(wait_timeout);
-                    return ::seastar::sleep(wait_timeout).then([] {
-                        return make_ready_future<std::optional<int>>();
-                    });
-                } else if (ret.result > 0) {
-                    return make_ready_future<std::optional<int>>(wstatus);
-                } else {
-                    ret.throw_if_error();
-                    return make_ready_future<std::optional<int>>(-1);
-                }
-            });
-        });
-    });
-}
-
 future<int> reactor::waitpid(pid_t pid) {
-    return _thread_pool->submit<syscall_result<int>>([pid] {
+    syscall_result<int> pidfd = co_await _thread_pool->submit<syscall_result<int>>([pid] {
         return wrap_syscall<int>(syscall(__NR_pidfd_open, pid, O_NONBLOCK));
-    }).then([pid, this] (syscall_result<int> pidfd) {
-        if (pidfd.result == -1) {
-            // pidfd_open() was introduced in linux 5.3, so the pidfd.error could be ENOSYS on
-            // older kernels. But it could be other error like EMFILE or ENFILE. anyway, we
-            // should always waitpid().
-            return do_waitpid(pid);
-        } else {
-            return do_with(pollable_fd(file_desc::from_fd(pidfd.result)), [pid, this](auto& pidfd) {
-                return pidfd.readable().then([pid, this] {
-                    return do_waitpid(pid);
-                });
-            });
-        }
     });
+    // pidfd_open() was introduced in linux 5.3, so the pidfd.error could be ENOSYS on
+    // older kernels. But it could be other error like EMFILE or ENFILE. anyway, we
+    // should always waitpid().
+    std::optional<pollable_fd> pfd;
+    if (pidfd.result != -1) {
+        pfd.emplace(file_desc::from_fd(pidfd.result));
+        co_await pfd->readable();
+    }
+
+    std::chrono::milliseconds wait_timeout(0);
+    for (;;) {
```

Comment:
Please don't code things as infinite loops, it means the reader has to reverse-engineer the termination condition.

Reply:
Fixed in [ee292c4](https://github.com/scylladb/seastar/commit/ee292c4d00de19a9104049f23d3d66ae3e82bbae)